### PR TITLE
Fix #15 : Implement closing method for websocket

### LINKED ISSUE #15: Add close method to Socket client
At the moment there is a maxMessages variable to trigger when the socket run ends. I don't think this is overly user friendly, how would I know what to set this to and how would I gracefully shut down the socket if I need to exit early.

Instead we should add a Close method to the socket client. This can be called to shutdown and end the socket. 

I'm not 100% familiar with the gorilla websocket, but the close method would need to shutdown the gorilla socket and end the run loop and clear our any processing messages.  You may need to employ a notify channel to exit the Run loop, I have an example of this here https://github.com/libsv/bitcoin-hc/blob/master/transports/zmq/headers.go#L60

Doing this means the caller can control when the Run loop ends and removes the need to a user to know in advance how many messages they want to process.

### PR DESCRIPTION
**Pull request by ctnguyen**

### CODE REVIEW

File: test/integration/Websocket_test.go [hunk 1]
```diff
@@ -48,30 +49,29 @@ func getChannelTokens() (string, string, string, error) {
...
+	totalRecv := uint64(0)
+
+	ws := spv.NewWSClient(
+		spv.WithBaseURL(baseURL),
+		spv.WithVersion(version),
+		spv.WithToken(token1),
+		spv.WithChannelID(channelid),
+		spv.WithWebsocketCallBack(func(ctx context.Context, t int, msg []byte, err error) error {
+			atomic.AddUint64(&totalRecv, 1)
+			//fmt.Print(string(msg))
```
Comment (roppa) [inline test/integration/Websocket_test.go:62]: comment left in

File: Client.go [hunk 2]
```diff
@@ -251,9 +272,9 @@ func (c *Client) sendRequest(req *http.Request, out interface{}) error {
...
 type WSClient struct {
-	cfg        *spvConfig
-	ws         *ws.Conn
-	nbNotified uint64
+	cfg   *spvConfig
+	ws    *ws.Conn
+	close chan bool
 }
 
 // NewWSClient create a new websocket client by providing fuctional config settings.
```
Comment (roppa) [inline Client.go:280]: typo should be functional

File: example/websocket/main.go [hunk 3]
```diff
@@ -2,17 +2,20 @@ package main
...
 
-var channelid = "b1j-Vd94XrU9NJlnrtQPfkzOgFLxun5oWZLUhXfvnZk2cekKe4QY7YKh_hbXivAroApDtVn3pmmOo848R6BhAw"
-var tok = "hqaDcOY-3svYqZv5RXID7AphKp9Bm8obQ_74K7mFLjcjq_Bw-Vwng6Q0q7PvJqhKawikfmd0Kr2OpYFKpFrKcg"
+var channelid = "cc33DR4-U1ZJnvKvMChUcikxck5ANY1XhIbkIz5YPXJIZCS--mDIUc9Ot5HbLRSBIZuFFtNZzKIMz8StV46-cw"
+var tok = "T_udxsz-1sE9RPeBmEgYNNTtFaQEW204ETf3DzcpwQydIz_Gvb7X3gB6rPypkQmIH5Fl9_cfiZdk6XD3uzTmoQ"
 
 // PullUnreadMessages pull the notified unread messages, mark them as read
-func PullUnreadMessages(t int, msg []byte, err error) error {
+func PullUnreadMessages(ctx context.Context, t int, msg []byte, err error) error {
+	//ctx, cancelfn := context.WithTimeout(ctx, time.Second)
```
Comment (roppa) [inline example/websocket/main.go:17]: Ctx is commented, was this left in by mistake?

**Review by roppa — COMMENTED** [review]
Just some commented out code and a typo

File: client.go [hunk 4]
```diff
@@ -296,45 +321,36 @@ type WSClient struct {
...
-
-// Run establishes the connection and start listening the notification stream
-// process the notification if a callback is provided
-func (c *WSClient) Run() error {
+// connectServer establish the connection to the server
+// Return error if any
+func (c *WSClient) connectServer() error {
 	u := url.URL{
 		Scheme: "wss",
 		Host:   "localhost:5010",
```
Comment (theflyingcodr) [inline client.go:356]: Should these values should be read from the config?

**Review by theflyingcodr — CHANGES_REQUESTED** [review]
Just 1 thing I spotted, should the ws connection read from the cfg rather than hardcoded localhost etc?

File: client.go (hunk 4 above)
Comment (ctnguyen) [reply client.go:356]: oh nice one. I'll fix that

**Pull request merged**